plugins/plugin-pullrequest/codenvy-plugin-pullrequest-ide/src/main/java/com/codenvy/plugin/pullrequest/client/ContributionMixinProvider.java

```diff
@@ -0,0 +1,135 @@
+import com.codenvy.plugin.pullrequest.client.vcs.VcsServiceProvider;
+import com.codenvy.plugin.pullrequest.client.vcs.hosting.VcsHostingService;
+import com.codenvy.plugin.pullrequest.client.vcs.hosting.VcsHostingServiceProvider;
+import com.codenvy.plugin.pullrequest.client.workflow.WorkflowExecutor;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+import org.eclipse.che.api.promises.client.Function;
+import org.eclipse.che.api.promises.client.FunctionException;
+import org.eclipse.che.api.promises.client.Operation;
+import org.eclipse.che.api.promises.client.OperationException;
+import org.eclipse.che.api.promises.client.Promise;
+import org.eclipse.che.api.promises.client.PromiseError;
+import org.eclipse.che.api.promises.client.js.Promises;
+import org.eclipse.che.ide.api.project.MutableProjectConfig;
+import org.eclipse.che.ide.api.resources.Project;
+import org.eclipse.che.ide.api.resources.Resource;
+import org.eclipse.che.ide.api.resources.ResourceInterceptor;
+
+import static com.codenvy.plugin.pullrequest.projecttype.shared.ContributionProjectTypeConstants.CONTRIBUTE_TO_BRANCH_VARIABLE_NAME;
+import static com.codenvy.plugin.pullrequest.projecttype.shared.ContributionProjectTypeConstants.CONTRIBUTION_PROJECT_TYPE_ID;
+import static java.util.Collections.singletonList;
+
+/**
+ * Resource interceptor, which provides contribution mixin to the project, which loads from the server.
+ *
+ * @author Vlad Zhukovskyi
+ * @see ResourceInterceptor
+ * @since 5.0.0
+ */
+@Singleton
+public class ContributionMixinProvider implements ResourceInterceptor {
+
+    private VcsServiceProvider        vcsServiceProvider;
+    private VcsHostingServiceProvider vcsHostingServiceProvider;
+    private WorkflowExecutor          workflowExecutor;
+    private ContributePartPresenter   contributionPartPresenter;
+
+    @Inject
+    public ContributionMixinProvider(VcsServiceProvider vcsServiceProvider,
+                                     VcsHostingServiceProvider vcsHostingServiceProvider,
+                                     WorkflowExecutor workflowExecutor,
+                                     ContributePartPresenter contributionPartPresenter) {
+        this.vcsServiceProvider = vcsServiceProvider;
+        this.vcsHostingServiceProvider = vcsHostingServiceProvider;
+        this.workflowExecutor = workflowExecutor;
+        this.contributionPartPresenter = contributionPartPresenter;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void intercept(Resource resource) {
+        if (!resource.isProject()) {
+            return;
+        }
+
+        if (resource.getLocation().segmentCount() != 1) {
+            return;
+        }
+
+        if (vcsServiceProvider.getVcsService((Project)resource) == null) {
+            return;
+        }
+
+        final Project project = (Project)resource;
+
+        if (project.getMixins().contains(CONTRIBUTION_PROJECT_TYPE_ID)) {
+            return;
+        }
+
```

Comment:
Then I guess you did not succeeded, because such conditions as `resource.getLocation().segmentCount() != 1` are not really informative enough. For readability reasons I would like to recommend extracting those conditions into correspondingly named private methods and then combining those into a single condition. Something like this:
```java 
if (resourceIsNotAProject() ||
    resourceIsNotARoot() ||
    resourceHasNoVcsService ||
    resourceIsNotAContribution) {
    return;
}
``` 
But generally speaking there should not be so many checks for pre-conditions. Usually such cases are a sign of not accurately planned class architecture, however I admit that there always can be exceptions :-)